aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp

```diff
@@ -0,0 +1,178 @@
+  // On empty input, no output data will be generated,
+  // so use arbitrary qparams.
+  float x_min = 0;
+  float x_max = 0;
+  // Otherwise...
+  if (input.numel() > 0) {
+    x_min = input.min().item<float>();
+    x_max = input.max().item<float>();
+  }
+
+  // Input tensor is quantized as 8-bit unsigned values
+  static constexpr int precision = 8;
+  static constexpr bool is_signed = false;
+
+  // Calculate scale and zero point for quantization of input tensor
+  auto q_params = quant_utils::ChooseQuantizationParams(
+      /*min=*/x_min,
+      /*max=*/x_max,
+      /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+      /*qmax=*/
+      is_signed ? ((1 << (precision - 1)) - 1) : (1 << precision) - 1,
+      /*preserve_sparsity=*/false,
+      /*force_scale_power_of_two=*/false,
+      /*reduce_range=*/reduce_range);
+
+  // Quantize input
+  at::Tensor q_input = at::quantize_per_tensor(
+      input, q_params.scale, q_params.zero_point, c10::kQUInt8);
+
+  at::Tensor out =
+      apply_impl<false>(q_input, q_params.scale, q_params.zero_point);
+
+  return at::dequantize(out); // TODO: optimized kernel that outputs fp32 so
+                              // this step isn't necessary
+}
+
+template at::Tensor PackedConvWeightsQnnp<2>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range);
+
+template at::Tensor PackedConvWeightsQnnp<3>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range);
+
+#endif // USE_PYTORCH_QNNPACK
+
+namespace at {
+namespace native {
+namespace {
+
+// note: this works for both Conv and ConvT due to transpose()
+template <int kSpatialDim>
+class QConvDynamicInt8 final {
+ public:
+  static at::Tensor run(
+      at::Tensor input,
+      const c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>>&
+          packed_weight,
+      bool reduce_range) {
+    return packed_weight->apply_dynamic(input, reduce_range);
+  }
+};
+
+// note: this works for both Conv and ConvT due to transpose()
+class QConv1dDynamicInt8 final {
+ public:
+  static at::Tensor run(
+      at::Tensor input,
+      const c10::intrusive_ptr<ConvPackedParamsBase<2>>& packed_weight,
+      bool reduce_range) {
+    at::Tensor output;
+    // N, C, L -> N, C, 1, L
+    input = input.unsqueeze(quant_utils::kConv1dSqueezeDim + 2);
+    output = packed_weight->apply_dynamic(input, reduce_range);
+    // N, C, 1, L -> N, C, L
+    return output.squeeze_(quant_utils::kConv1dSqueezeDim + 2);
+  }
+};
+
+TORCH_LIBRARY_IMPL(quantized, CPU, m) {
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::conv1d_dynamic"),
+      TORCH_FN(QConv1dDynamicInt8::run));
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::conv2d_dynamic"),
+      TORCH_FN(QConvDynamicInt8<2>::run));
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::conv3d_dynamic"),
+      TORCH_FN(QConvDynamicInt8<3>::run));
+
+  // transpose
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::conv_transpose1d_dynamic"),
+      TORCH_FN(QConv1dDynamicInt8::run));
```

Comment:
why is conv transpose using the exact same class as conv?

Reply:
I dont remember the details and @z-a-f added it probably. But for the most part it is likely because they share much of the functionality.

Reply:
yes, its the same as for the statically quantized ops